File: python/equity/01-options-pricing/tests/test_binomial.py
"""CRR binomial tree: convergence to BS, American exercise properties."""

import math

import numpy as np
import pytest

from eq_options import bs_price, crr_price, early_exercise_premium

CASES = [
    # (S, K, T, r, sigma, q)
    (100.0, 100.0, 1.0, 0.05, 0.20, 0.00),
    (100.0, 110.0, 0.5, 0.03, 0.30, 0.02),
    (100.0, 80.0, 2.0, 0.01, 0.15, 0.04),
    (50.0, 60.0, 0.25, -0.01, 0.40, 0.00),
]


@pytest.mark.parametrize(("S", "K", "T", "r", "sigma", "q"), CASES)
@pytest.mark.parametrize("otype", ["call", "put"])
def test_european_tree_converges_to_bs_500_steps(
    S: float, K: float, T: float, r: float, sigma: float, q: float, otype: str
) -> None:
    bs = bs_price(S, K, T, r, sigma, q, otype)
    tree = crr_price(S, K, T, r, sigma, q, otype, "european", 500)
    assert tree == pytest.approx(bs, abs=2e-2, rel=2e-3)


def test_european_tree_tight_at_5000_steps() -> None:
    bs = bs_price(100, 100, 1, 0.05, 0.2, 0.02, "call")
    tree = crr_price(100, 100, 1, 0.05, 0.2, 0.02, "call", "european", 5000)
    assert tree == pytest.approx(bs, abs=2e-3)


def test_monotone_convergence_sanity() -> None:
    """|error| decreases as steps double (same parity kills CRR odd/even
    oscillation); error * n stays bounded (O(1/n) rate)."""
    S, K, T, r, sigma, q = 100.0, 105.0, 1.0, 0.04, 0.25, 0.01
    bs = bs_price(S, K, T, r, sigma, q, "call")
    steps = [50, 100, 200, 400, 800]
    errs = [
        abs(crr_price(S, K, T, r, sigma, q, "call", "european", n) - bs)
        for n in steps
    ]
    for e_coarse, e_fine in zip(errs, errs[1:]):
        assert e_fine < e_coarse
    # O(1/n): error at 800 steps should be ~16x smaller than at 50 steps
    assert errs[-1] < errs[0] / 8.0


def test_crr_convergence_rate_fitted_exponent() -> None:
    """Fit the CRR discretisation-error exponent by log-log regression.

    CRR is a first-order scheme: error(n) ~ C / n, with an odd/even (and,
    away from at-the-money, node-alignment) oscillation superposed on the
    leading term. A single pair of step counts is not enough to *prove*
    the O(1/n) rate -- a two-point ratio can land anywhere depending on
    where each n happens to fall in the oscillation. Fitting a line
    through log|error| vs log(n) over a decade of geometrically spaced
    step counts averages the oscillation out and recovers the leading
    exponent, which must come out close to the theoretical -1.
    """
    S, K, T, r, sigma, q = 100.0, 105.0, 1.0, 0.04, 0.25, 0.01
    bs = bs_price(S, K, T, r, sigma, q, "call")
    steps = np.array([200 * 2**k for k in range(9)])  # 200 .. 51200
    errs = np.array(
        [
            abs(crr_price(S, K, T, r, sigma, q, "call", "european", int(n)) - bs)
            for n in steps
        ]
    )
    assert np.all(errs > 0.0), "tree exactly matches BS at some n -- fit is degenerate"
    slope, _intercept = np.polyfit(np.log(steps), np.log(errs), 1)
    assert -1.3 < slope < -0.7, (
        f"fitted CRR convergence exponent {slope:.3f}; theory predicts -1 "
        "(error ~ C/n) -- this is more than a loose oscillation, investigate"
    )


@pytest.mark.parametrize(("S", "K", "T", "r", "sigma", "q"), CASES)
def test_american_put_geq_european_put(
    S: float, K: float, T: float, r: float, sigma: float, q: float
) -> None:
    amer = crr_price(S, K, T, r, sigma, q, "put", "american", 400)
    euro = crr_price(S, K, T, r, sigma, q, "put", "european", 400)
    assert amer >= euro - 1e-12


def test_american_call_no_dividend_equals_european() -> None:
    """With q=0 early exercise of a call is never optimal (Merton)."""
    for r in (0.0, 0.05, 0.10):
        amer = crr_price(100, 100, 1, r, 0.25, 0.0, "call", "american", 500)
        euro = crr_price(100, 100, 1, r, 0.25, 0.0, "call", "european", 500)
        assert amer == pytest.approx(euro, abs=1e-10)


def test_american_call_with_dividend_carries_premium() -> None:
    premium = early_exercise_premium(100, 90, 2.0, 0.02, 0.2, 0.06, "call", 500)
    assert premium > 1e-3


@pytest.mark.parametrize(("S", "K", "T", "r", "sigma", "q"), CASES)
@pytest.mark.parametrize("otype", ["call", "put"])
def test_early_exercise_premium_nonnegative(
    S: float, K: float, T: float, r: float, sigma: float, q: float, otype: str
) -> None:
    assert early_exercise_premium(S, K, T, r, sigma, q, otype, 200) >= 0.0


def test_american_put_at_least_intrinsic() -> None:
    """Deep ITM American put must be worth at least immediate exercise."""
    for S in (40.0, 60.0, 80.0):
        amer = crr_price(S, 100.0, 1.0, 0.08, 0.2, 0.0, "put", "american", 400)
        assert amer >= (100.0 - S) - 1e-10


def test_deep_itm_american_put_near_intrinsic_high_rates() -> None:
    """High carry makes early exercise almost certain: value ~ intrinsic."""
    amer = crr_price(20.0, 100.0, 1.0, 0.10, 0.15, 0.0, "put", "american", 400)
    assert amer == pytest.approx(80.0, abs=1e-6)


def test_famous_american_put_benchmark() -> None:
    """S=K=100, T=1, r=5%, sigma=20%: American put ~ 6.0896 (literature)."""
    amer = crr_price(100, 100, 1.0, 0.05, 0.20, 0.0, "put", "american", 2000)
    assert amer == pytest.approx(6.0896, abs=5e-3)


def test_tree_put_call_parity_european() -> None:
    """European tree prices satisfy parity to tree accuracy."""
    S, K, T, r, sigma, q, n = 100.0, 95.0, 1.5, 0.03, 0.25, 0.02, 800
    c = crr_price(S, K, T, r, sigma, q, "call", "european", n)
    p = crr_price(S, K, T, r, sigma, q, "put", "european", n)
    assert c - p == pytest.approx(
        S * math.exp(-q * T) - K * math.exp(-r * T), abs=1e-9
    )


def test_single_step_tree_by_hand() -> None:
    """n=1 CRR reproduces the hand-computed one-step value."""
    S, K, T, r, sigma = 100.0, 100.0, 1.0, 0.05, 0.2
    u = math.exp(sigma)
    d = 1 / u
    p = (math.exp(r) - d) / (u - d)
    expected = math.exp(-r) * p * (S * u - K)
    assert crr_price(S, K, T, r, sigma, 0.0, "call", "european", 1) == pytest.approx(
        expected, abs=1e-12
    )


def test_invalid_exercise_and_steps_raise() -> None:
    with pytest.raises(ValueError, match="exercise"):
        crr_price(100, 100, 1, 0.05, 0.2, 0.0, "call", "bermudan", 100)  # type: ignore[arg-type]
    with pytest.raises(ValueError, match="n_steps"):
        crr_price(100, 100, 1, 0.05, 0.2, 0.0, "call", "european", 0)

File: cpp/equity-options-engine/tests/test_binomial.cpp
// CRR binomial tree: convergence to Black-Scholes, American/European
// ordering, early-exercise identities, edge cases and input validation.

#include <gtest/gtest.h>

#include <cmath>
#include <numeric>
#include <stdexcept>
#include <vector>

#include "eqopt/binomial.hpp"
#include "eqopt/black_scholes.hpp"

namespace {

using eqopt::bs_price;
using eqopt::crr_price;
using eqopt::early_exercise_premium;
using eqopt::ExerciseStyle;
using eqopt::OptionType;

TEST(BinomialConvergence, EuropeanWithin2e3At2000Steps) {
    const double S = 100.0, K = 100.0, T = 1.0, r = 0.05, q = 0.02,
                 sigma = 0.2;
    for (OptionType type : {OptionType::Call, OptionType::Put}) {
        const double bs = bs_price(S, K, T, r, sigma, q, type);
        const double tree = crr_price(S, K, T, r, sigma, q, type,
                                      ExerciseStyle::European, 2000);
        EXPECT_NEAR(tree, bs, 2e-3);
    }
}

TEST(BinomialConvergence, ErrorShrinksWithSteps) {
    const double S = 100.0, K = 110.0, T = 0.75, r = 0.03, q = 0.01,
                 sigma = 0.25;
    const double bs = bs_price(S, K, T, r, sigma, q, OptionType::Call);
    const auto err = [&](int n) {
        return std::abs(crr_price(S, K, T, r, sigma, q, OptionType::Call,
                                  ExerciseStyle::European, n) -
                        bs);
    };
    // Average consecutive odd/even trees to kill the oscillating term, then
    // require monotone decay across decades of n.
    const auto smooth_err = [&](int n) { return 0.5 * (err(n) + err(n + 1)); };
    const double e50 = smooth_err(50);
    const double e200 = smooth_err(200);
    const double e800 = smooth_err(800);
    const double e2000 = smooth_err(2000);
    EXPECT_LT(e200, e50);
    EXPECT_LT(e800, e200);
    EXPECT_LT(e2000, e800);
    EXPECT_LT(e2000, 2e-3);
}

TEST(BinomialConvergence, FittedExponentMatchesTheoreticalOrderOne) {
    // CRR is a first-order scheme: error(n) ~ C / n, with an odd/even (and
    // node-alignment) oscillation superposed on the leading term. A single
    // pair of step counts (as above) cannot *prove* the O(1/n) rate -- a
    // two-point ratio can land anywhere in the oscillation. Regressing
    // log|error| on log(n) over a decade of geometrically spaced step
    // counts averages the oscillation out and recovers the leading
    // exponent, which must come out close to the theoretical -1.
    const double S = 100.0, K = 105.0, T = 1.0, r = 0.04, q = 0.01,
                 sigma = 0.25;
    const double bs = bs_price(S, K, T, r, sigma, q, OptionType::Call);
    const std::vector<int> steps{200, 400, 800, 1600, 3200, 6400, 12800, 25600, 51200};
    std::vector<double> log_n, log_err;
    for (int n : steps) {
        const double err = std::abs(crr_price(S, K, T, r, sigma, q,
                                              OptionType::Call,
                                              ExerciseStyle::European, n) -
                                    bs);
        ASSERT_GT(err, 0.0) << "tree exactly matches BS at n=" << n
                            << " -- fit is degenerate";
        log_n.push_back(std::log(static_cast<double>(n)));
        log_err.push_back(std::log(err));
    }
    // Ordinary least squares slope of log_err ~ slope * log_n + intercept.
    const double mean_x =
        std::accumulate(log_n.begin(), log_n.end(), 0.0) / log_n.size();
    const double mean_y =
        std::accumulate(log_err.begin(), log_err.end(), 0.0) / log_err.size();
    double num = 0.0, den = 0.0;
    for (size_t i = 0; i < log_n.size(); ++i) {
        num += (log_n[i] - mean_x) * (log_err[i] - mean_y);
        den += (log_n[i] - mean_x) * (log_n[i] - mean_x);
    }
    const double slope = num / den;
    EXPECT_GT(slope, -1.3) << "fitted CRR exponent " << slope
                           << "; theory predicts -1 (error ~ C/n)";
    EXPECT_LT(slope, -0.7) << "fitted CRR exponent " << slope
                           << "; theory predicts -1 (error ~ C/n)";
}

TEST(BinomialAmerican, AmericanAtLeastEuropeanEverywhere) {
    const std::vector<double> strikes{80.0, 100.0, 120.0};
    const std::vector<double> vols{0.1, 0.3};
    const std::vector<double> divs{0.0, 0.04};
    for (double K : strikes)
        for (double sigma : vols)
            for (double q : divs)
                for (OptionType type : {OptionType::Call, OptionType::Put}) {
                    const double amer = crr_price(100.0, K, 1.0, 0.05, sigma,
                                                  q, type,
                                                  ExerciseStyle::American, 400);
                    const double euro = crr_price(100.0, K, 1.0, 0.05, sigma,
                                                  q, type,
                                                  ExerciseStyle::European, 400);
                    EXPECT_GE(amer, euro - 1e-12)
                        << "K=" << K << " sigma=" << sigma << " q=" << q;
                }
}

TEST(BinomialAmerican, NoDividendAmericanCallEqualsEuropean) {
    // With q = 0 early exercise of a call is never optimal: the American and
    // European prices coincide on the same tree (exact node-by-node).
    for (double K : {80.0, 100.0, 120.0}) {
        const double amer = crr_price(100.0, K, 1.0, 0.05, 0.25, 0.0,
                                      OptionType::Call,
                                      ExerciseStyle::American, 500);
        const double euro = crr_price(100.0, K, 1.0, 0.05, 0.25, 0.0,
                                      OptionType::Call,
                                      ExerciseStyle::European, 500);
        EXPECT_NEAR(amer, euro, 1e-12);
        EXPECT_DOUBLE_EQ(
            early_exercise_premium(100.0, K, 1.0, 0.05, 0.25, 0.0,
                                   OptionType::Call, 500),
            0.0);
    }
}

TEST(BinomialAmerican, PutCarriesEarlyExercisePremiumAndDominatesIntrinsic) {
    // Deep ITM American put: value >= intrinsic, and premium > 0 when r > q.
    const double amer = crr_price(80.0, 100.0, 1.0, 0.05, 0.2, 0.0,
                                  OptionType::Put, ExerciseStyle::American,
                                  500);
    EXPECT_GE(amer, 20.0 - 1e-12);
    EXPECT_GT(early_exercise_premium(80.0, 100.0, 1.0, 0.05, 0.2, 0.0,
                                     OptionType::Put, 500),
              0.0);
}

TEST(BinomialEdgeCases, DegenerateLimitsMatchPolicy) {
    // T == 0 -> intrinsic.
    EXPECT_DOUBLE_EQ(crr_price(110.0, 100.0, 0.0, 0.05, 0.2, 0.0,
                               OptionType::Call, ExerciseStyle::American, 100),
                     10.0);
    // sigma == 0, European -> Black-Scholes deterministic limit.
    EXPECT_DOUBLE_EQ(
        crr_price(100.0, 95.0, 1.0, 0.05, 0.0, 0.02, OptionType::Call,
                  ExerciseStyle::European, 100),
        bs_price(100.0, 95.0, 1.0, 0.05, 0.0, 0.02, OptionType::Call));
    // sigma == 0, American put with r > 0: exercising now beats waiting.
    const double amer0 = crr_price(80.0, 100.0, 1.0, 0.05, 0.0, 0.0,
                                   OptionType::Put, ExerciseStyle::American,
                                   100);
    EXPECT_NEAR(amer0, 20.0, 1e-12);
    // S == 0 American put pays K immediately.
    EXPECT_DOUBLE_EQ(crr_price(0.0, 100.0, 1.0, 0.05, 0.2, 0.0,
                               OptionType::Put, ExerciseStyle::American, 100),
                     100.0);
    // K == 0 put is worthless.
    EXPECT_DOUBLE_EQ(crr_price(100.0, 0.0, 1.0, 0.05, 0.2, 0.0,
                               OptionType::Put, ExerciseStyle::European, 100),
                     0.0);
}

TEST(BinomialValidation, InvalidInputsThrow) {
    EXPECT_THROW(crr_price(-1.0, 100.0, 1.0, 0.05, 0.2), std::invalid_argument);
    EXPECT_THROW(crr_price(100.0, 100.0, 1.0, 0.05, 0.2, 0.0,
                           OptionType::Call, ExerciseStyle::European, 0),
                 std::invalid_argument);
    // dt too large for r - q vs sigma: risk-neutral probability leaves (0,1).
    EXPECT_THROW(crr_price(100.0, 100.0, 1.0, 5.0, 0.01, 0.0,
                           OptionType::Call, ExerciseStyle::European, 1),
                 std::invalid_argument);
}

}  // namespace
